### PR TITLE
[rocprofv3] Change file output in tool_detach to occur on a separate thread so tool_detach will return immediately

### PR DESCRIPTION
**Pull request by itrowbri**
## Motivation

Customer made a code change suggestion where file output generation in `tool_detach` is changed to occur on a separate, detached thread. This allows `tool_detach` to immediately return, fixing a hang/termination issue when profiling large vLLMs like DeepSeek where detaching would cause the workload to terminate due to how long it was taking to generate the requested output files. 

## Technical Details

The function `generate_output` is now run on a detached thread after `tool_detach` is called.

## Test Plan

Existing attach/detach tests should pass.

## Submission Checklist

- [x] Look over the contributing guidelines at https://github.com/ROCm/ROCm/blob/develop/CONTRIBUTING.md#pull-requests.

### CODE REVIEW

File: projects/rocprofiler-sdk/source/docs/how-to/using-rocprofv3-process-attachment.rst [hunk 1]
```diff
@@ -114,6 +114,7 @@ Reattaching to the same process
 The dynamic process attachment functionality supports reattachment, which can be used to attach multiple times to the same PID over a process's lifetime. Give the same PID to ``rocprofv3`` to reattach.
 
 There are some restrictions on what options can change when reattaching.  Typically, tracing, PC sampling, ATT, counter collection, and other options that change what data will be collected cannot be changed. ``rocprofv3`` will throw a ``RuntimeError`` if it detects a configuration change that is not supported.
+Furthermore, output file generation from the previous attachment must be finished. Otherwise, errors or data corruption might occur after the profiler is reattached.
```
Comment (meserve-amd) [inline projects/rocprofiler-sdk/source/docs/how-to/using-rocprofv3-process-attachment.rst:117]: Minor: add newline before

File: projects/rocprofiler-sdk/source/lib/rocprofiler-sdk-tool/tool.cpp [hunk 2]
```diff
@@ -2840,7 +2840,9 @@ tool_detach(void* /*tool_data*/)
     if(tool_metadata->process_end_ns == 0)
         rocprofiler_get_timestamp(&(tool_metadata->process_end_ns));
 
-    generate_output(cleanup_mode::reset);
+    // Launch generate output in a seperate thread and detach it
+    // This allows tool_detach to return immediately without waiting
+    std::thread([]() { generate_output(cleanup_mode::reset); }).detach();
```
Comment (bwelton) [inline projects/rocprofiler-sdk/source/lib/rocprofiler-sdk-tool/tool.cpp]: Isn't the issue that the application itself terminates, wouldn't that in turn kill this thread as well?

[the remark this answers is not in the document]
Comment (meserve-amd) [reply projects/rocprofiler-sdk/source/lib/rocprofiler-sdk-tool/tool.cpp]: I think we should follow up on this, at the very least we should protect against a cleanup/destroy operation in fini

**Review by meserve-amd — COMMENTED** [review]
Looks ok, please address the copilot comments as well

File: projects/rocprofiler-sdk/source/lib/rocprofiler-sdk-tool/tool.cpp (hunk 2 above)
Comment (itrowbri) [reply projects/rocprofiler-sdk/source/lib/rocprofiler-sdk-tool/tool.cpp]: The problem, from what I understand, is not the program shutting down, but that the output generation takes so long the application thinks the worker process is hanging and terminates it. Detaching from a DeepSeek benchmark can result in over 10 minutes of output generation, and trying to have an active worker process output files before returning to work often results in the benchmark terminating early due to a perceived hang.

I'm not 100% sure about what happens if the main program tries to shutdown while the output thread is still working. I asked about it, but I'm not 100% sure they understood what I was asking. I can run some more tests and ask for some more details if that's the main concern you're highlighting here.

[the remark this answers is not in the document]
Comment (bwelton) [reply projects/rocprofiler-sdk/source/lib/rocprofiler-sdk-tool/tool.cpp]: Yeah there is also a concern here about what happens if the application terminates while this thread is processing.

**Comment by itrowbri** [comment]
Adding instructions to reproduce error that occurs **without** this PR change:

Start a rocm:vllm container, e.g.
```
docker run -it --network=host --group-add=video --ipc=host --cap-add=SYS_PTRACE \
--security-opt seccomp=unconfined --privileged --device /dev/kfd --device /dev/dri \
-v /home/<user>:/dockerx -v /data:/data --workdir /dockerx -d --name <docker_name> \
 rocm/vllm:latest
```

Set up a vllm server in another CLI for the same container (or run in the background):
```

ROCP_TOOL_ATTACH=1  vllm serve /data/DeepSeek-R1 \
        --served-model-name DeepSeek-R1 \
        --trust-remote-code \
        -tp 8 \
        --no-enable-chunked-prefill \
        --disable-log-requests \
        --max-model-len 32768 \
        --block-size 1 \
        --gpu_memory_utilization 0.95
```

Start the benchmark on another CLI in the same container
```
vllm bench serve     --model deepseek-ai/DeepSeek-R1    --served-model-name DeepSeek-R1    --random-input-len 1     --random-output-len 300     --num-prompts 100     --max-concurrency 50     --ignore-eos --profile
```

Attach to a worker process on another CLI in the same container:
```
rocprofv3 -r --output-format pftrace --log-level info --attach <pid> | tee profiler.log
```

Attach for several minutes or as long as you can before the benchmark ends. Before the benchmark ends, detach. Without the change in this PR, the benchmark should timeout as the worker process attempts to output data.

File: projects/rocprofiler-sdk/source/lib/rocprofiler-sdk-tool/tool.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (ammarwa) [reply projects/rocprofiler-sdk/source/lib/rocprofiler-sdk-tool/tool.cpp]: I would create a thread and save it, then join it after the rocprof finishes all the work. For future-proofing, I recommend allocating a static pool of threads created in the init, and using one from it whenever we need a new thread. We ensure we join all these threads and delete them at finalization.

**Review by ammarwa — CHANGES_REQUESTED** [review]

File: projects/rocprofiler-sdk/source/docs/how-to/using-rocprofv3-process-attachment.rst (hunk 1 above)
Comment (ammarwa) [reply projects/rocprofiler-sdk/source/docs/how-to/using-rocprofv3-process-attachment.rst:117]: I don't think we need that note anymore with the new changes I added to the `tool.cpp`

File: projects/rocprofiler-sdk/source/lib/rocprofiler-sdk-tool/tool.cpp [hunk 3]
```diff
@@ -2840,7 +2844,18 @@ tool_detach(void* /*tool_data*/)
...
+    // This allows tool_detach to return immediately without waiting
+    {
+        std::lock_guard<std::mutex> lock(::cleanup_thread_mutex);
+        if(::cleanup_thread)
+        {
+            if(::cleanup_thread->joinable()) ::cleanup_thread->join();
+            ::cleanup_thread.reset();
+        }
+        ::cleanup_thread =
+            std::make_unique<std::thread>([]() { generate_output(cleanup_mode::reset); });
```
[the remark this answers is not in the document]
Comment (ammarwa) [reply projects/rocprofiler-sdk/source/lib/rocprofiler-sdk-tool/tool.cpp]: detach is unsafe.

[the remark this answers is not in the document]
Comment (ammarwa) [reply projects/rocprofiler-sdk/source/lib/rocprofiler-sdk-tool/tool.cpp]: The mutex lock prevents enqueuing multiple generations of output, and the thread join ensures that the current thread waits for the previous thread to finish.

File: projects/rocprofiler-sdk/source/lib/rocprofiler-sdk-tool/tool.cpp [hunk 4]
```diff
@@ -125,6 +125,10 @@ rocprofv3_error_signal_handler(int signo, siginfo_t*, void*);
 
 namespace
 {
+// Thread for safe cleanup output generation
```
[the remark this answers is not in the document]
Comment (ammarwa) [reply projects/rocprofiler-sdk/source/lib/rocprofiler-sdk-tool/tool.cpp:129]: Yes, there is room for improvement, but the comment suggestion here is not good.

File: projects/rocprofiler-sdk/source/lib/rocprofiler-sdk-tool/tool.cpp [hunk 5]
```diff
@@ -2840,7 +2844,18 @@ tool_detach(void* /*tool_data*/)
...
+    {
+        std::lock_guard<std::mutex> lock(::cleanup_thread_mutex);
+        if(::cleanup_thread)
+        {
+            if(::cleanup_thread->joinable()) ::cleanup_thread->join();
+            ::cleanup_thread.reset();
+        }
+        ::cleanup_thread =
+            std::make_unique<std::thread>([]() { generate_output(cleanup_mode::reset); });
+    }
```
Comment (ammarwa) [inline projects/rocprofiler-sdk/source/lib/rocprofiler-sdk-tool/tool.cpp]:
```suggestion
        if(!::cleanup_thread || !::cleanup_thread->joinable())
        {
            ::cleanup_thread =
                std::make_unique<std::thread>([]() { generate_output(cleanup_mode::reset); });
        }
    }
```

Another room for improvement is to skip `generate_output` when another active thread is running, since the active thread will handle these results.

File: projects/rocprofiler-sdk/source/lib/rocprofiler-sdk-tool/tool.cpp [hunk 6]
```diff
@@ -125,6 +125,10 @@ rocprofv3_error_signal_handler(int signo, siginfo_t*, void*);
 
 namespace
 {
+// Thread for safe cleanup output generation
+std::unique_ptr<std::thread> cleanup_thread = nullptr;
+std::mutex                   cleanup_thread_mutex;
```
Comment (jrmadsen) [inline projects/rocprofiler-sdk/source/lib/rocprofiler-sdk-tool/tool.cpp]:
```suggestion
auto output_generation_thread = Synchronized<std::unique_ptr<std::thread>>{};
```

File: projects/rocprofiler-sdk/source/lib/rocprofiler-sdk-tool/tool.cpp [hunk 7]
```diff
@@ -2840,7 +2844,18 @@ tool_detach(void* /*tool_data*/)
...
-    generate_output(cleanup_mode::reset);
+    // Launch generate output in a separate thread for safe cleanup
+    // This allows tool_detach to return immediately without waiting
+    {
+        std::lock_guard<std::mutex> lock(::cleanup_thread_mutex);
+        if(::cleanup_thread)
+        {
+            if(::cleanup_thread->joinable()) ::cleanup_thread->join();
+            ::cleanup_thread.reset();
+        }
```
Comment (jrmadsen) [inline projects/rocprofiler-sdk/source/lib/rocprofiler-sdk-tool/tool.cpp]: You need to join the thread during tool attachment.

File: projects/rocprofiler-sdk/source/lib/rocprofiler-sdk-tool/tool.cpp [hunk 8]
```diff
@@ -2853,6 +2868,16 @@ tool_fini(void* /*tool_data*/)
...
+    // Join the cleanup thread if it exists and is active
+    {
+        std::lock_guard<std::mutex> lock(::cleanup_thread_mutex);
+        if(::cleanup_thread)
+        {
+            if(::cleanup_thread->joinable()) ::cleanup_thread->join();
+            ::cleanup_thread.reset();
+        }
+    }
+
```
Comment (jrmadsen) [inline projects/rocprofiler-sdk/source/lib/rocprofiler-sdk-tool/tool.cpp:2882]: This is duplicate code from above. Move to a function.

**Review by jrmadsen — CHANGES_REQUESTED** [review]

File: projects/rocprofiler-sdk/source/lib/rocprofiler-sdk-tool/tool.cpp (hunk 3 above)
Comment (jrmadsen) [inline projects/rocprofiler-sdk/source/lib/rocprofiler-sdk-tool/tool.cpp]:
```suggestion
            std::make_unique<std::thread>(generate_output, cleanup_mode::reset);
```

File: projects/rocprofiler-sdk/source/lib/rocprofiler-sdk-tool/tool.cpp (hunk 2 above)
Comment (jrmadsen) [reply projects/rocprofiler-sdk/source/lib/rocprofiler-sdk-tool/tool.cpp]: It looks like you resolved this but the issue they failed to account for is if the application terminates while output generation is happening because, yes, if that happens, this thread will be killed.

File: projects/rocprofiler-sdk/source/lib/rocprofiler-sdk-tool/tool.cpp (hunk 5 above)
Comment (jrmadsen) [reply projects/rocprofiler-sdk/source/lib/rocprofiler-sdk-tool/tool.cpp]: > Another room for improvement is to skip `generate_output` when another active thread is running, since the active thread will handle these results.

This would be quite dangerous. We shouldn't have a thread generating output while data is being collected, our data structures at present could not handle that.
